clang/include/clang/Basic/DiagnosticSemaKinds.td

```diff
@@ -13603,6 +13603,19 @@ def note_unsafe_count_attributed_pointer_argument : Note<
 def warn_unsafe_single_pointer_argument : Warning<
   "unsafe assignment to function parameter of __single pointer type">,
   InGroup<UnsafeBufferUsage>, DefaultIgnore;
+def warn_unsafe_assign_to_null_terminated_pointer : Warning<
+  "unsafe %select{assignment to a pointer of|"
+                  "assignment to a parameter of|"
+                  "return of|"
+                  "conversion to|"
+                  "initialization of a pointer of|"
+                  "assignment to a pointer of|"
+                  "casting to}2 '__null_terminated' type; "
+  "only '__null_terminated' pointers, string literals, and 'std::string::c_str' calls are compatible with '__null_terminated' pointers">,
```

Comment:
It might be simpler to say that. the source pointer may not point to null-terminated data.  I could imagine that we eventually we will increase the number of cases that we allow.  I don't have super strong opinions about this - it is nice to tell people what is allowed, but it doesn't tell them what the error is.

Reply:
I see. Yeah, I should explain what the error is.   The convention for warning messages (at least for -Wunsafe-buffer-usage) is that we have a short but clear enough sentence to just say "something is unsafe", then put necessary explanation in notes.   My message here might already be lengthy.  I will explain what the error is in a note and also move  `"only '__null_terminated' pointers, ..."` (sort of the suggestion) to a note too.

Reply:
oh, I maybe misunderstood part of your comment here.  
You were suggesting to simply saying "source pointer may not point to null-terminated data." instead of breaking into cases.    I'm not against the idea of making the warning message simpler.  My only concern is that warning messages under `-Wunsafe-buffer-usage` all have the forms "unsafe [some operation]" or "[some operation] is unsafe".  I'd like this message to follow that.

Different cases are determined by `AssignmentAction`, which is handled by the existing framework in `Sema`.  I assume it is stable.

Reply:
Or maybe `"unsafe conversion to __null_terminated type; source pointer may not point to null-terminated data."` ?  If "conversion" is a appropriate umbrella term for these operations.